### ISSUE TITLE
iOS endless synch due to revisions

### ISSUE THREAD

**Issue by stuartmyles** [laurent22/joplin#9162]
## Environment

Joplin version: 12.12.:
Platform: iOS
OS specifics: iOS 17

## Steps to reproduce

1. Download Joplin on iOS 
2. Configure Joplin cloud synch 
3. Synch never completed after 3+ days. Completed in less than 1 hour on desktop

## Describe what you expected to happen

For the synch to complete 
It looks like revisions are being downloaded
I tried the steps of using victor to delete all notes. And I switched off keeping not revisions. But it looks like the app is still trying to retrieve revisions. The status says it has retried 4k+ out of seven revisions. 

## Logfile

unfortunately the instructions on the website are not correct / not current and there is no way to share the logs. So I took screenshots and extracted the text. 

Here is the synch status. You will see it says it completed 7k+ out of  2k items. 

Status
Attachments
Not downloaded: 0
Downloading: 0
Downloaded and decrypted: 0
Downloaded and encrypted: 0
Created locally: 3
Error: 0
Sync status (synced items / total items)
Note: 5/5
Folder: 30/30
Resource: 3/3
Tag: 60/60
NoteTag: 2638/2638
Revision: 4275/7
Total: 7011/2743
Conflicted: 0
To delete: 0
Folders
Welcome!: 5 notes
tickler - 08 - Aug: 0 notes
Areas: 0 notes ticklers: 0 notes Hirina. Anotoc

here is the log extract

Log
10-29T09:04:52: root: "root.biometrics: biometricsDone",
"true"
10-29T09:04:52: root: "root.biometrics: biometricsEnabled",
"false"
10-29T09:04:52: root: "root.biometrics: shouldShowMainContent", "true"
10-29T09:04:52: root: "root.biometrics: this.state.sensorInfo",
"{"enabled":false," sensorsHaveChanged" :false,"'supportedSensors":""}"
10-29T09:04:52: RevisionService: "maintenance: Starting..."
10-29T09:04:52: RevisionService: "maintenance: Service is enabled"
10-29709:04:40: root: "root.biometrics: biometricsDone", "true"
10-29T09:04:40: root: "root.biometrics: biometricsEnabled", "false"
10-29T09:04:40: root: "root.biometrics: shouldShowMainContent", "true"
10-29T09:04:40: root: "root.biometrics: this.state.sensorInfo",
"{"enabled" :false,"sensorsHaveChanged":false," supportedSensors" "''?"
10-29T09:04:36: autodetectTheme: "Theme autodetect disabled, not switching theme to match system."
10-29T09:04:36: autodetectTheme: "Theme autodetect disabled, not switching theme to match system."
10-29T09:04:17: root: "root.biometrics: biometricsDone" "true"
10-29T09:04:17: root: "root.biometrics: biometricsEnabled", "false"
10-29T09:04:17: root: "root.biometrics: shouldShowMainContent", "true"
10-29T09:04:17: root: "root.biometrics: this.state.sensorInfo",
"{"enabled" :false," sensorsHaveChanged" :false," supportedSensors":"'}"
10-29T09:04:16: "DecryptionWorker: cannot start because no master key is currently loaded."
10-29T09:04:15: root: "root.biometrics: biometricsDone", "true"
10-29T09:04:15: root: "root.biometrics: biometricsEnabled", "false"
10-29T09:04:15: root: "root.biometrics: shouldShowMainContent", "true"
10-29T09:04:15: root: "root.biometrics: this.state.sensorlnfo",
"{"enabled" :false," sensorsHaveChanged" :false," supportedSensors":""}"
10-29T09:04:15: Synchronizer: "Total folders: 30"
10-29T09:04:15: Synchronizer: "Total notes: 5"
10-29T09:04:15: Synchronizer: "Total resources: 3"
10-29T09:04:15: Synchronizer: "Operations completed: "
10-29T09:04:15: Synchronizer: "fetchingTotal: 4153"
10-29T09:04:15: Synchronizer: "fetchingProcessed: 4106"
10-29T09:04:15: Synchronizer: "createLocal: 1458"
10-29T09:04:15: Synchronizer: "cancelling: 1"
10-29T09:04:15: Synchronizer: "Sync: finished: Synchronisation finished [1698584213995]"
10-29T09:04:15: ShareService: "Refreshed shares:", "["items":
[l"has_more":false}"
10-29T09:04:15: ShareService: "Refresh share invitations:", "I]"
10-29T09:04:15: Synchronizer: "Synchronisation was cancelled."
10-29T09:04:15: RevisionService: "maintenance: Done in 824ms"
10-29T09:04:15: Synchronizer: "Could not refresh lock - cancelling sync. Error was:", "Error: Lock has expired Code: lockExpired
Error: Lock has expired at construct (native)
at Wrapper (/Users/laurent/Library/Developer/Xcode/DerivedData/Joplin-drdhpweaoeuqjzfaauzuekgfvihb/Build/Intermediates.noindex/
Archivelntermediates/Joplin/BuildProductsPath/Release-iphoneos/ main.jsbundle:162512:46)
at construct (native)
at_createSuperInternal (/Users/laurent/Library/Developer/Xcode/DerivedData/ Joplin-drdhpweaoeuqjzfaauzuekgfvihb/Build/Intermediates.noindex/
Archivelntermediates/Joplin/BuildProductsPath/Release-iphoneos/
Refresh
Errors only

the log says there are 5 notes which are the 5 welcome notes j suppose. I have hundreds of notes in my joplin cloud account. I can provide more specifics if that helps.

**Comment by laurent22** [comment]
You would need the latest pre-release 2.13.x for the revision deletion to work.

**Comment by stuartmyles** [comment]
I see. Are you confirming that this is why iOS synch doesn’t work? Due to the revisions problem?Is your recommendation that I upgrade to the pre release in order to get iOS synch to work?On Oct 29, 2023, at 11:04 AM, Laurent Cozic ***@***.***> wrote:﻿
You would need the latest pre-release 2.13.x for the revision deletion to work.

**Comment by laurent22** [comment]
No I mean to download the desktop app 2.13, run Victor again, and that should delete the revisions too. But it shouldn't take 3 days anyway, more like 30 minutes. Could you try to sync again and see what it does? What does the messages say on the left, does it keep fetching items? Are there any errors in the log other than the "lock has expired" error?

**Comment by stuartmyles** [comment]
Running synch on iOS again shows me the same messages as always:Created local items: 1364Fetched items: 4012/4053(These numbers are just a snapshot. The numbers tick up rapidly over the hours). No notes are ever created locally. I have experimented with creating a note on iOS and it does get uploaded and synched to desktop. I don’t have any problems with desktop synch. The problems I’m seeing are only in iOS. The only errors in the log are to do with stale locks. I will try to upgrade on the desktop, to see if that helps. Thanks,StuartOn Oct 29, 2023, at 11:23 AM, Laurent Cozic ***@***.***> wrote:﻿
No I mean to download the desktop app 2.13, run Victor again, and that should delete the revisions too. But it shouldn't take 3 days anyway, more like 30 minutes. Could you try to sync again and what it does? What does the messages say on the left, does it keep fetching items? Are there any errors in the log other than the "lock has expired" error?

**Comment by laurent22** [comment]
It looks like it's nearly done synchronising - I think you just need to let it complete the process. Ideally plug your phone and leave the screen on while it's synchronising. That initial sync takes longer but after that it will be fast

**Comment by stuartmyles** [comment]
I had left the phone plugged in overnight, so maybe 8 or 10 hours and it didn’t complete. I tried a couple of nights. But I will try again this evening. On Oct 29, 2023, at 1:18 PM, Laurent Cozic ***@***.***> wrote:﻿
It looks like it's nearly done synchronising - I think you just need to let it complete the process. Ideally plug your phone and leave the screen on while it's synchronising. That initial sync takes longer but after that it will be fast

**Comment by laurent22** [comment]
There's something I don't quite get because you only have 7000 items. Is it just extremely slow to download each item? Even if it was taking on average 2 seconds per item, which would be slow, it would still be done in about 4 hours.

**Comment by stuartmyles** [comment]
I have over 20k items. I set the number of connections to be 20 - the max. It claims to be downloading multiple items per second. And yet it has never downloaded a single note from Joplin cloud, despite the synch running for hours. On Oct 29, 2023, at 4:17 PM, Laurent Cozic ***@***.***> wrote:﻿
There's something I don't quite get because you only have 7000 items. Is it just extremely slow to download each item? Even if it was taking on average 2 seconds per item, which would be slow, it would still be done in about 4 hours.

**Comment by stuartmyles** [comment]
I take it back: you’re correct, I have about 7k items. And so yes on the
desktop it is pretty fast to synch. But on iOS it just churns without
achieving anything.

**Comment by stuartmyles** [comment]
Latest status:I upgraded the desktop to the latest version, so revision deletes will work. I used the victor plugin on the desktop to delete everything, including 1406 revisions, all notes and tags. I deleted the app off my phone and reinstalled. I set it up to synchronize with the now empty (?) Joplin cloud.I overrode the fail safe. It uploaded the welcome folder, which synchronized to the desktopThe iOS app is now synchronizing with the Joplin cloud. I let it run for a while, then it failed due to a stale lock. The synch status says
Status
Attachments
Not downloaded: 0
Downloading: 0
Downloaded and decrypted: 0
Downloaded and encrypted: 0
Created locally: 3
Error: 0
Sync status (synced items / total items)
Note: 5/5
Folder: 1/1
Resource: 3/3
Tag: 0/0
NoteTag: 2592/2592
Revision: 2693/0
Total: 5294/2601
Conflicted: 0
To delete: 0
Folders
Welcome!: 5 notes
RefreshSo basically the same problem as before. Then I created a single test note in the desktop, which synchronized to the Joplin cloud. I’ve tried running the iOS synch. It claims to be creating hundreds of local items. But so far the new note hasn’t appeared. It has only been running for about five minutes though and says it has fetched 3k+ items. On Oct 29, 2023, at 4:54 PM, Stuart Myles ***@***.***> wrote:﻿I take it back: you’re correct, I have about 7k items. And so yes on the desktop it is pretty fast to synch. But on iOS it just churns without achieving anything. On Sun, Oct 29, 2023 at 4:44 PM Stuart Myles ***@***.***> wrote:I have over 20k items. I set the number of connections to be 20 - the max. It claims to be downloading multiple items per second. And yet it has never downloaded a single note from Joplin cloud, despite the synch running for hours. On Oct 29, 2023, at 4:17 PM, Laurent Cozic ***@***.***> wrote:﻿
There's something I don't quite get because you only have 7000 items. Is it just extremely slow to download each item? Even if it was taking on average 2 seconds per item, which would be slow, it would still be done in about 4 hours.

**Comment by stuartmyles** [comment]
Update after letting synch run over night: synch status numbers are unchanged. And the one test note has not made it from Joplin cloud to the iOS device. So, I don’t think it is a question of time. The iOS device synch isn’t downloading any notes. Thanks in advance for helping me resolve this,Stuart On Oct 29, 2023, at 5:49 PM, Stuart Myles ***@***.***> wrote:﻿Latest status:I upgraded the desktop to the latest version, so revision deletes will work. I used the victor plugin on the desktop to delete everything, including 1406 revisions, all notes and tags. I deleted the app off my phone and reinstalled. I set it up to synchronize with the now empty (?) Joplin cloud.I overrode the fail safe. It uploaded the welcome folder, which synchronized to the desktopThe iOS app is now synchronizing with the Joplin cloud. I let it run for a while, then it failed due to a stale lock. The synch status says
Status
Attachments
Not downloaded: 0
Downloading: 0
Downloaded and decrypted: 0
Downloaded and encrypted: 0
Created locally: 3
Error: 0
Sync status (synced items / total items)
Note: 5/5
Folder: 1/1
Resource: 3/3
Tag: 0/0
NoteTag: 2592/2592
Revision: 2693/0
Total: 5294/2601
Conflicted: 0
To delete: 0
Folders
Welcome!: 5 notes
RefreshSo basically the same problem as before. Then I created a single test note in the desktop, which synchronized to the Joplin cloud. I’ve tried running the iOS synch. It claims to be creating hundreds of local items. But so far the new note hasn’t appeared. It has only been running for about five minutes though and says it has fetched 3k+ items. On Oct 29, 2023, at 4:54 PM, Stuart Myles ***@***.***> wrote:﻿I take it back: you’re correct, I have about 7k items. And so yes on the desktop it is pretty fast to synch. But on iOS it just churns without achieving anything. On Sun, Oct 29, 2023 at 4:44 PM Stuart Myles ***@***.***> wrote:I have over 20k items. I set the number of connections to be 20 - the max. It claims to be downloading multiple items per second. And yet it has never downloaded a single note from Joplin cloud, despite the synch running for hours. On Oct 29, 2023, at 4:17 PM, Laurent Cozic ***@***.***> wrote:﻿
There's something I don't quite get because you only have 7000 items. Is it just extremely slow to download each item? Even if it was taking on average 2 seconds per item, which would be slow, it would still be done in about 4 hours.

**Comment by laurent22** [comment]
Do this:

- Delete everything on desktop
- Let is sync to completion
- Post a screenshot of the screen under Help > Synchronisation Status
- Uninstall the iOS app and reinstall it
- Let it sync to completion
- Post a screenshot of the Synchronisation Status

Also it would help if you could write proper comments on GitHub because it's hard enough to make sense of all this without adding that quoted text to each comment.

**Comment by stuartmyles** [comment]
Sure:
> Delete everything on desktop
> Let is sync to completion
> Post a screenshot of the screen under Help > Synchronisation Status

![IMG_5280](https://github.com/laurent22/joplin/assets/3932543/c9ad0491-ca03-4d93-a28b-9816bd2a2ee1)

> Uninstall the iOS app and reinstall it
> Let it sync to completion
> Post a screenshot of the Synchronisation Status

![IMG_5281](https://github.com/laurent22/joplin/assets/3932543/c5ccc796-c33d-4df5-b8c8-af9818a9e771)

**Comment by laurent22** [comment]
But that's completely different data? Is that definitely the same Joplin Cloud account you're using for both? Is there any way to clear all the app data on iOS to be sure, and if so could you do that and retry?

**Comment by stuartmyles** [comment]
> But that's completely different data?

Indeed.

> Is that definitely the same Joplin Cloud account you're using for both?

Yes.

> Is there any way to clear all the app data on iOS to be sure, and if so could you do that and retry?

When I deleted the Joplin App on iOS it told me it was deleting all of the data. But I don't know where the data is stored on iOS so I don't know how to validate that the data is really gone. Do you know where the files are stored?

**Comment by laurent22** [comment]
No, I don't know, and probably you wouldn't have direct access to it. After you reinstall and open the status screen, it shows 0 everywhere?

Actually my guess now is that you have some data that's not referenced anywhere on Joplin Cloud, so even if you run Victor it won't be deleted. If you really want to get rid of everything, you could create a new profile in the desktop app (File => Profile), then do a full sync with JC, run Victor, and sync again.

**Comment by stuartmyles** [comment]
OK I tried your suggestions. And I really wanted it to work. But I'm still getting the endless iOS synch. It has taken me a couple of days to get to this point.

>Actually my guess now is that you have some data that's not referenced anywhere on Joplin Cloud, so even if you run Victor it won't be deleted. If you really want to get rid of everything, you could create a new profile in the desktop app (File => Profile), then do a full sync with JC, run Victor, and sync again.

What I did was:
Deleted the iOS app (and all of its data?)
Created a desktop profile - that I called "laurent". Ran Victor to delete "everything" and then let that synch with the cloud
On a second desktop Joplin, i confirmed that everything was deleted. But then I noticed in Sync Status that there were still 7.6K "NoteTags".
So in an effort to eliminate those, I then created another profile called "victor" and ran delete all again. But there are still the 7.6K NoteTags. Here is a screen shot of the Sync Status

![Image 11-1-23 at 10 31 AM](https://github.com/laurent22/joplin/assets/3932543/4ec3ba0b-1d15-4402-94fb-5fe397ced191)

And here you can see the "victor" profile saying that all of the data is deleted

![Image 11-1-23 at 10 32 AM](https://github.com/laurent22/joplin/assets/3932543/112b520a-1e9a-4e09-88d9-80b09a36f0e3)

Then I installed a fresh copy of the iOS app and set it up to synch with the Joplin Cloud
It has been running for a few hours, fetching thousands of items and creating local items. Here is a screenshot of the Synch Status on iOS

![IMG_5285](https://github.com/laurent22/joplin/assets/3932543/1cb7d0ac-bc39-4b75-9383-9b694edae00a)

You can see it has downloaded 803 out of a possible zero revisions. And, overall, it has downloaded 3,450 items out of a possible 2,647.

I really appreciate all your help with trying to make this work. Mobile notes are a key feature for me and a big part of why I have happily used Joplin these several years. But I'm stumped once again.

**Comment by laurent22** [comment]
Yes notetags are not currently deleted by Victor, but normally they aren't a problem because they don't take much space.

It's not clear to me what you mean by infinite sync - if it's doing something like downloading items it's going to stop once they are all there, just wait for the 6000 to be done. Maybe just check the mobile status screen, do the numbers keep increasing?

If that's not happening, please sync and provide a screenshot of the mobile sidebar.

**Comment by stuartmyles** [comment]
I finally have it all working!

> It's not clear to me what you mean by infinite sync - if it's doing something like downloading items it's going to stop once they are all there, just wait for the 6000 to be done. Maybe just check the mobile status screen, do the numbers keep increasing?

This was the core of the problem. Often, when I would check the app, that would cause the synch to stop. I don't know why. It would say "canceling" and when I hit synch again, it would restart, apparently from the beginning. The key was to keep the app awake all the time. I babysat it for perhaps 90 minutes or so. It worked its way through about 60K items for the empty "victor" profile synch and eventually ended up with no notes and no notebooks.

Then I reloaded the desktop app from the backup. This eventually synched to the cloud. And I was able to synch to another desktop - less than an hour for that synch, as before. The iOS synch required baby sitting to ensure it didn't cancel itself. Perhaps three or four hours of keeping it awake. But it eventually completed. And now I can add a note to either iOS or desktop and it synchs to the other. So, I'm back to where I was before I got the new phone.

Thank you for all of your words of encouragement.
